sklearn/utils/testing.py

```diff
@@ -764,21 +765,29 @@ def _delete_folder(folder_path, warn=False):
 
 class TempMemmap(object):
     def __init__(self, data, mmap_mode='r'):
-        self.temp_folder = tempfile.mkdtemp(prefix='sklearn_testing_')
         self.mmap_mode = mmap_mode
         self.data = data
 
     def __enter__(self):
-        fpath = op.join(self.temp_folder, 'data.pkl')
-        joblib.dump(self.data, fpath)
-        data_read_only = joblib.load(fpath, mmap_mode=self.mmap_mode)
-        atexit.register(lambda: _delete_folder(self.temp_folder, warn=True))
+        data_read_only, self.temp_folder = create_memmap_backed_data(
+            self.data, mmap_mode=self.mmap_mode, return_folder=True)
         return data_read_only
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         _delete_folder(self.temp_folder)
 
 
+def create_memmap_backed_data(data, mmap_mode='r', return_folder=False):
+    temp_folder = tempfile.mkdtemp(prefix='sklearn_testing_')
+    atexit.register(functools.partial(_delete_folder, temp_folder, warn=True))
+    filename = op.join(temp_folder, 'data.pkl')
+    joblib.dump(data, filename)
```

Comment:
Stupid question, but it's not possible to create a memap without physically writing to disk? Creating temp files in multiple `check_estimators_*` functions (for each estimator in scikit-learn) doesn't seem ideal - that certainly wouldn't help test run time.  Joblib seems to [explicitly forbid this](https://github.com/joblib/joblib/blob/e7491c87ef4083d8064deb796f0ad1ad128f3462/joblib/test/test_numpy_pickle.py#L710) and this also doesn't work with `np.load(..., mmap_mode='r')` but I wonder if there is a workaround. We don't need an actual mmap, just something like behaves like it. 

Alternatively is it possible to cache the `create_memmap_backed_data` function? the data should be repeated multiple times for different estimators. 

Mostly a side comment, as this was already the case before this PR..

Reply:
> but it's not possible to create a memap without physically writing to disk? 

Looks like there is https://github.com/joblib/joblib/issues/378 , but it's out of the scope for this PR ..